### ISSUE TITLE
Dialogs aren't always positioned correctly; don't always have scrollbars when they should

### ISSUE THREAD

**Issue by Lange** [nodecg/nodecg#309]
## Expected outcome
Dialogs (both `paper-dialogs` that are part of the dashboard itself and NodeCG dialogs that bundles create) should always be centered in the viewport and have a scrollbar if their content does not fit in the viewport.

## Actual outcome
Dialogs aren't always centered and sometimes their content goes way off the bottom of the screen. Below is an example screenshot of the `vaadin-upload` dialog on the Assets page misbehaving:
![image](https://user-images.githubusercontent.com/873012/26981048-9fe05712-4cf9-11e7-8303-e9f0dc716747.png)

## Suggested action
Figure out why the dialogs are doing this and fix it. Test all `paper-dialogs` on the dashboard and also test NodeCG bundle dialogs to ensure that they behave as expected.

**Comment by Lange** [comment]
In the case of the specific dialog used in the example, part of the problem is that the content of this dialog was not wrapped in a `<paper-dialog-scrollable>` element. Adding that does give a scrollbar, but the positioning of the dialog is still incorrect:
![image](https://user-images.githubusercontent.com/873012/26981308-5c4a6424-4cfa-11e7-9f75-8ef4c3485c45.png)

**Comment by Lange** [comment]
Looks like the remainder of the problem is twofold:
1. The `fitInto` property of a `paper-dialog` tells the dialog what element it should try to fit itself into. By default, this element is `window`. This is no good for us, because we have this app toolbar that renders on top of the dialog. So, we need to ensure that we set the `fitInto` of every `paper-dialog` to a value that makes sense. In this particular case of the Assets upload dialog, I am setting it to the `<section>` that contains the Assets page. Here's how I'm doing that in `ncg-asset-category.js` (which is where this dialog lives):
    ```js
    connectedCallback() {
        super.connectedCallback();
        this.$.uploadDialog.fitInto = document.body.querySelector('ncg-dashboard').shadowRoot
            .querySelector('section[route="assets"]');
        this.$.uploadDialog.resetFit();
    }
    ```
2. However, setting `fitInto` alone is not enough. The `paper-dialog` must be informed of any changes in size to its content. It cannot automatically detect this and do the right thing. So it's up to us to call `.refit()` on the dialog whenever we know that its contents could have changed size.

**Comment by Lange** [comment]
The best way to summarize this issue is: back when I first added all these dialogs, I assumed that it "just worked". This is not the case, and a bit of work has to be done to ensure that dialogs always appear as intended.

**Comment by Lange** [comment]
Closed by 7025d51.
